### PR TITLE
fix(route_handler): route handler overlap removal is too conservative

### PR DESCRIPTION
**Pull request by mkquda**
## Description

Currently, overlap handling is not great, in some corner cases, a broken route might be generated, like the one shown below.

Screenshot from 2024-05-17 14-19-00

This is happening because, in the function getLaneletSequenceUpTo, if at least one of the previous lanelets is same as the inut lanelet, then it is considered a loop and search will stop.

Screenshot from 2024-05-17 14-19-00 (copy)

Another issue is the logic used to get the closest lanelet within route does not respect the history of the lanelets traversed, it will just return the closest route lanelet satisfying the distance and yaw constraints, so the closest lanelet can jump suddenly in case of partially overlapping lanelets causing a break in the route, as shown below.

https://github.com/autowarefoundation/autoware.universe/assets/168697710/307ae9eb-0bdb-4271-9ab9-ccec38bbd516

## Changes

**(route_handler)**
Modified the loop check logic in function getLaneletSequenceUpTo():
- If previous lanelets are route lanelets, then return true only if all prev lanelets are overlapping with input lanelet
- If at least one prev lanelet is not overlapping with input lanelet then continue sequence search
- If previous lanelets are not route lanelets and at least one overlaps with input lanelet then loop check returns true

Implemented new function getClosestRouteLaneletFromCurrent() to get closest route lanelet with respect the previous computed closest route lanelets to insure continuity.

**(planner_manager)**
Implemented separate function updateCurrentRouteLanelet() and moved logic out of getReferencePathFunction()
Use new function getClosestRouteLaneletFromCurrent() in updateCurrentRouteLanelet()

**(behavior_path_planner_common/utils)**
Modified function isEgoOutOfRoute() to take closestRoadLanelet as input from planner manager rather than recomputing same logic locally

**(mission_planner)**
Refactored DefaultPlanner::plan() function, should not affect logic.

## Tests performed

- Tested on psim
- Confirmed no degradation using autoware evaluator

## Effects on system behavior

Can generate proper route:
Screenshot from 2024-05-29 11-26-35

Closest route lanelet doesn't jump suddenly causing unexpected behavior:

https://github.com/autowarefoundation/autoware.universe/assets/168697710/a15067fd-79a4-4e3a-a557-d0412224340e

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/route_handler/src/route_handler.cpp [hunk 1]
```diff
@@ -880,8 +893,38 @@ bool RouteHandler::getClosestLaneletWithConstrainsWithinRoute(
     route_lanelets_, search_pose, closest_lanelet, dist_threshold, yaw_threshold);
 }
 
-bool RouteHandler::getNextLaneletWithinRoute(
-  const lanelet::ConstLanelet & lanelet, lanelet::ConstLanelet * next_lanelet) const
+boost::optional<lanelet::ConstLanelet> RouteHandler::getClosestRouteLaneletFromCurrent(
+  const Pose & search_pose, const lanelet::ConstLanelet & current_lanelet,
```
Comment (maxime-clem) [inline planning/route_handler/src/route_handler.cpp:897]: `current_lanelet` does not match the parameter name defined in the `.hpp` file.

File: planning/route_handler/src/route_handler.cpp [hunk 2]
```diff
@@ -880,8 +893,38 @@ bool RouteHandler::getClosestLaneletWithConstrainsWithinRoute(
...
+  boost::optional<lanelet::ConstLanelet> closest_lanelet = {};
+
+  lanelet::ConstLanelets previous_lanelets, next_lanelets, lanelet_sequence;
+  if (getPreviousLaneletsWithinRoute(current_lanelet, &previous_lanelets)) {
+    lanelet_sequence = previous_lanelets;
+  }
+
+  lanelet_sequence.push_back(current_lanelet);
+  const auto & current_neighbors =
+    lanelet::utils::query::getAllNeighbors(routing_graph_ptr_, current_lanelet);
```
Comment (maxime-clem) [inline planning/route_handler/src/route_handler.cpp:909]: Neighbor lanelets may not be in the route. Is this okay ?

Comment (mkquda) [reply planning/route_handler/src/route_handler.cpp:909]: you are right, will fix

File: planning/route_handler/include/route_handler/route_handler.hpp [hunk 3]
```diff
@@ -246,6 +248,22 @@ class RouteHandler
...
+   * lanelet.
+   * @param search_pose pose to find closest lanelet to
+   * @param reference_lanelet reference lanelet to decide the search set
+   * @param dist_threshold distance constraint closest lanelet must be within
+   * @param yaw_threshold yaw constraint closest lanelet direction must be within
+   */
+  bool getClosestRouteLaneletFromLanelet(
+    const Pose & search_pose, const lanelet::ConstLanelet & reference_lanelet,
+    lanelet::ConstLanelet * closest_lanelet, const double dist_threshold,
+    const double yaw_threshold) const;
```
Comment (zulfaqar-azmi-t4) [inline planning/route_handler/include/route_handler/route_handler.hpp:265]: possible to return `std::optional<lanelet::ConstLanelet>` instead?

Comment (mkquda) [reply planning/route_handler/include/route_handler/route_handler.hpp:265]: I did that first, but changed it for sake of consistency

**Review by zulfaqar-azmi-t4 — APPROVED** [review]
LGTM!

**Review by maxime-clem — APPROVED** [review]
Tested and it works well

**Review by kosuke55 — APPROVED** [review]
LGTM

**Pull request merged**